File: 12. coffre-fort/main.py
# https://prologin.org/train/2022/qualification/coffre-fort

# Énoncé

# Joseph Marchand doit ouvrir un coffre électronique sécurisé.

# Le coffre contient un circuit complexe, décrit sous la forme d'un ensemble de puces, reliées entre elles par des fils. Il existe toujours exactement un chemin entre chaque paire de puces. Chaque puce envoie un signal

# au verrou lorsqu'elle est activée. Joseph Marchand tente de passer outre la sécurité du coffre et a besoin de votre aide pour calculer rapidement les effets de ses manipulations.

# Il veut régulièrement connecter les deux bornes d'un générateur électrique à deux puces
# et du circuit, et se demande quel signal est envoyé au verrou du coffre suite à cette opération. Une puce est activée s'il existe un chemin de vers passant par , qui ne repasse pas deux fois par le même fil. Le signal reçu par le verrou est alors le produit du signal envoyé par chaque puce activée, modulo

# (car on ne peut pas représenter de trop grands nombres par un signal électrique).
# Exemple

# exemple
# Entrée

# L'entrée contiendra :

#     Sur la première ligne, un entier : N, nombre de puces.
#     Sur la ligne suivante, un entier : M, nombre de fils, égal à N-1
#     Sur la ligne suivante, un entier : R, nombre de questions.
#     Sur la ligne suivante, une liste de N entiers séparés par des espaces : signaux, liste des signaux.
#     Sur les lignes suivantes, une liste de M éléments : fils, liste des fils entre les puces.
#         Une ligne par élément de la liste : séparés par des espaces, un entier puce1 (première extrémité du fil), et un entier puce2 (seconde extrémité du fil).
#     Sur les lignes suivantes, une liste de R éléments : questions, liste des questions.
#         Une ligne par élément de la liste : séparés par des espaces, un entier puce a (première extrémité alimentée), et un entier puce b (seconde extrémité alimentée).

# Sortie

# Affiche le signal envoyé au coffre-fort pour chaque requête.
# Contraintes

# Runtime constraints

# Maximum memory usage
#     20000 kilobytes
# Maximum execution time
#     1000 milliseconds

# Input/output samples

# Sample input

#     6
#     5
#     5
#     4 9 7 2 3 0
#     0 1
#     1 3
#     1 2
#     2 4
#     4 5
#     0 3
#     1 3
#     0 4
#     0 5
#     1 2

# Sample output

#     72
#     18
#     756
#     0
#     63

# Sample input

#     12
#     11
#     5
#     1671404010 1671404009 1371404010 1 3 8 2 7 10 4 0 6
#     0 1
#     2 0
#     3 2
#     3 4
#     9 3
#     5 3
#     7 5
#     5 6
#     8 4
#     9 10
#     11 10
#     1 8
#     7 7
#     6 8
#     9 7
#     2 10

# Sample output

#     385444061
#     7
#     480
#     224
#     0

def calculer_signaux(n, m, r, signaux, fils, questions):
    """
    :param n: nombre de puces
    :type n: int
    :param m: nombre de fils
    :type m: int
    :param r: nombre de questions
    :type r: int
    :param signaux: liste des signaux
    :type signaux: list[int]
    :param fils: liste des fils entre les puces
    :type fils: list[dict["puce1": int, "puce2": int]]
    :param questions: liste des questions
    :type questions: list[dict["puce a": int, "puce b": int]]
    """
    # TODO Affiche le signal envoyé au coffre-fort pour chaque requête
    liaisons = {}
    for i in range(m):
        liaisons[i] = []
        for f in fils:
            if i in f.values():
                liaisons[i].append(f)
    for q in questions:
        # Obtention du chemin
        i = q['puce a']
        chemin = [i]
        faux_chemins = []
        parcouru = []
        while i != q['puce b']:
            fils_parcours = [
                x for x in liaisons[i] if x not in faux_chemins and x not in parcouru]
            for f in fils_parcours:
                i = f['puce2'] if i == f['puce1'] else f['puce1']
                chemin.append(i)
                dernier_f = f
                parcouru.append(f)
                break
            if len(fils_parcours) == 0:
                faux_chemins.append(dernier_f)
                i = q['puce a']
                chemin = [i]
                parcouru = []
        # Calcul et affichage
        p = 1
        for x in chemin:
            p *= signaux[x]
        print(p % 1671404011)
    pass


if __name__ == '__main__':
    n = int(input())
    m = int(input())
    r = int(input())
    signaux = list(map(int, input().split()))
    fils = [
        dict(zip(("puce1", "puce2"), map(int, input().split())))
        for _ in range(m)
    ]
    questions = [
        dict(zip(("puce a", "puce b"), map(int, input().split())))
        for _ in range(r)
    ]
    calculer_signaux(n, m, r, signaux, fils, questions)

File: 12. coffre-fort/main.cpp
// https://prologin.org/train/2022/qualification/coffre-fort
#include <algorithm>
#include <iostream>
#include <map>
#include <queue>
#include <vector>

struct Fil {
  int puce1;
  int puce2;
};

struct Question {
  int puce_a;
  int puce_b;
};

std::vector<int> puces_adj(std::vector<Fil> fils, int p) {
  std::vector<int> out;
  for (Fil f : fils) {
    if (f.puce1 == p)
      out.push_back(f.puce2);
    else if (f.puce2 == p)
      out.push_back(f.puce1);
  }
  return out;
}

std::vector<int> bfs(std::vector<Fil> fils, int a, int b) {
  // bfs
  std::vector<int> explored;
  std::queue<int> q;
  std::map<int, int> parents;
  explored.push_back(a);
  q.push(a);
  while (!q.empty()) {
    int v = q.front();
    q.pop();
    if (v == b)
      break;
    std::vector<int> adj = puces_adj(fils, v);
    for (int i : adj) {
      if (std::find(explored.begin(), explored.end(), i) == explored.end()) {
        explored.push_back(i);
        parents[i] = v;
        q.push(i);
      }
    }
  }
  // retracer le chemin à partir des parents
  std::vector<int> chemin = {b};
  int i = b;
  while (i != a) {
    chemin.push_back(parents[i]);
    i = parents[i];
  }
  return chemin;
}

std::vector<int> search(std::vector<Fil> Fils, int a, int b) {
  std::vector<int> explored;
  std::map<int, int> parents;
  std::deque<int> q;
  explored.push_back(a);
  q.push_back(a);
  while (!q.empty()) {
    int v = q.back();
    q.pop_back();
    if (v == b)
      break;
    for (auto i : puces_adj(Fils, v)) {
      if (std::find(explored.begin(), explored.end(), i) == explored.end()) {
        explored.push_back(i);
        q.push_back(i);
        parents[i] = v;
      }
    }
  }
  std::vector<int> chemin = {b};
  int i = b;
  while (i != a) {
    chemin.push_back(parents[i]);
    i = parents[i];
  }
  return chemin;
}

void calcul_signaux(int n, int m, int r, const std::vector<int> &signaux,
                    const std::vector<Fil> &fils,
                    const std::vector<Question> &questions) {
  for (Question q : questions) {
    auto sea = bfs(fils, q.puce_a, q.puce_b);
    /* for (int i : sea) */
    /*   std::cout << i << " "; */
    /* std::cout << std::endl; */
    int s = 1;
    for (int i : sea)
      s *= signaux[i];
    std::cout << s << std::endl;
    std::cout << (s % 1671404011) << std::endl;
  }
}

int main() {
  int n, m, r;
  std::cin >> n >> m >> r;

  std::vector<int> signaux(n);
  std::vector<Fil> fils(m);
  std::vector<Question> questions(r);
  for (int &i : signaux)
    std::cin >> i;
  for (Fil &i : fils)
    std::cin >> i.puce1 >> i.puce2;
  for (Question &i : questions)
    std::cin >> i.puce_a >> i.puce_b;

  calcul_signaux(n, m, r, signaux, fils, questions);
}
